mne/beamformer/_ap.py

```diff
@@ -0,0 +1,765 @@
+
+    Initialization: search the 1st source location over
+    the entire dipoles topographies space.
+
+    Parameters
+    ----------
+    attr_dict : dict
+        See: _calculate_free_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+    oris : array, shape (nsources, 3)
+        Orientations array of estimated sources (sorted by s_ap).
+    sub_g_proj : array
+        Sub space projected by estimated dipoles.
+
+    """
+    from scipy.linalg import eig
+
+    s_ap = []
+    oris = np.empty((attr_dict['nsources'], 3))
+    ap_val1 = np.zeros(attr_dict['ndipoles'])
+    perpend_spc = np.eye(gain.shape[0])
+    for dip in range(attr_dict['ndipoles']):
+        sol_tuple = _solve_active_gain_eig(dip, gain,
+                                           data_cov, eig,
+                                           perpend_spc)
+        ap_val1[dip] = np.max([x.real for x in sol_tuple[0]])
+
+    # obtain the 1st source location
+    s1_idx = np.argmax(ap_val1)
+    s_ap.append(s1_idx)
+
+    # obtain the 1st source orientation
+    sol_tuple = _solve_active_gain_eig(s1_idx, gain,
+                                       data_cov, eig,
+                                       perpend_spc)
+    oris[0] = sol_tuple[1][:,
+                           [np.argmax([x.real for x
+                                      in sol_tuple[0]])]
+                           ][:, 0]
+    sub_g_proj = np.dot(sol_tuple[2], oris[0])[:, np.newaxis]
+    return s_ap, oris, sub_g_proj
+
+
+def _free_phase1b(attr_dict, gain, data_cov,
+                  ap_temp_tuple, force_no_rep):
+    """Calculate phase 1b of free oriented AP.
+
+    Adding one source at a time.
+
+    Parameters
+    ----------
+    ap_temp_tuple : tuple
+        See: _free_phase1a.
+    attr_dict : dict
+        See: _calculate_free_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    force_no_rep : bool
+        Forces no repetition of estinated dipoles.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+    oris : array, shape (nsources, 3)
+        Orientations array of estimated sources (sorted by s_ap).
+    sub_g_proj : array
+        Sub space projected by estimated dipoles.
+
+    """
+    # ap_temp_tuple = (s_ap, oris, sub_g_proj)
+    from scipy.linalg import eig
+
+    s_ap, oris, sub_g_proj = copy(ap_temp_tuple)
+    for src in range(1, attr_dict['nsources']):
+        ap_val2 = np.zeros(attr_dict['ndipoles'])
+        act_spc = multi_dot([sub_g_proj,
+                            pinv(np.matmul(sub_g_proj.transpose(),
+                                           sub_g_proj)
+                                 ),
+                            sub_g_proj.transpose()])
```

Comment:
```suggestion
        act_spc = multi_dot([sub_g_proj, pinv(sub_g_proj.T @ sub_g_proj),
                             sub_g_proj.T])
```

This is another pattern I've seen a few times, might also consider a small helper function for it (`multi_dot([a, pinv(a.T @ a), a.T])`)